aws-java-sdk-core/src/main/java/com/amazonaws/http/conn/ssl/SdkTLSSocketFactory.java

```diff
@@ -130,26 +133,36 @@ public Socket connectSocket(
     private void verifyMasterSecret(final Socket sock) {
         if (sock instanceof SSLSocket) {
             SSLSocket ssl = (SSLSocket)sock;
-            SSLSession session = ssl.getSession();
+            final SSLSession session = ssl.getSession();
             if (session != null) {
-                String className = session.getClass().getName();
+                final String className = session.getClass().getName();
                 if ("sun.security.ssl.SSLSessionImpl".equals(className)) {
+                    boolean success = false;
+                    Object masterSecret = null;
                     try {
-                        Class<?> clazz = Class.forName(className);
-                        Method method = clazz.getDeclaredMethod("getMasterSecret");
-                        method.setAccessible(true);
-                        Object masterSecret = method.invoke(session);
-                        if (masterSecret == null)
-                            throw log(new SecurityException("Invalid SSL master secret"));
-                    } catch (ClassNotFoundException e) {
+                        masterSecret = AccessController.doPrivileged(new PrivilegedExceptionAction<Object>() {
+                          @Override
+                          public Object run() throws Exception {
+                            Class<?> clazz = Class.forName(className);
+                            Method method = clazz.getDeclaredMethod("getMasterSecret");
+                            method.setAccessible(true);
+                            return method.invoke(session);
+                          }
+                        });
+                        success = true;
+                    } catch (SecurityException e) {
                         failedToVerifyMasterSecret(e);
-                    } catch (NoSuchMethodException e) {
-                        failedToVerifyMasterSecret(e);
-                    } catch (IllegalAccessException e) {
-                        failedToVerifyMasterSecret(e);
-                    } catch (InvocationTargetException e) {
-                        failedToVerifyMasterSecret(e.getCause());
+                    } catch (PrivilegedActionException e) {
+                        // unwrap checked exception
+                        Throwable cause = e.getCause();
+                        if (cause instanceof InvocationTargetException) {
+                          // unwrap checked exception
+                          cause = cause.getCause();
+                        }
+                        failedToVerifyMasterSecret(cause);
                     }
+                    if (success && masterSecret == null)
```

Comment:
Thanks @SalusaSecondus for the suggestion (which I personally concur).

In any case, this pull request is currently under security review, and will likely be merged with other more recent changes introduced to this part of the code once we get the green light from the security team.